src/mlpack/methods/ann/layer/bilinear_interpolation_impl.hpp

```diff
@@ -0,0 +1,173 @@
+template<typename eT>
+void BilinearInterpolation<InputDataType, OutputDataType>::Forward(
+    const arma::Mat<eT>&& input, arma::Mat<eT>&& output)
+{
+  if (output.is_empty())
+    output.set_size(outRowSize * outColSize * depth, 1);
+
+  assert(input.n_rows == inRowSize * inColSize * depth);
+  assert(input.n_cols == 1);
+  assert(output.n_rows == outRowSize * outColSize * depth);
+  assert(output.n_cols == 1);
+  assert(inRowSize >= 2);
+  assert(inColSize >= 2);
+
+  arma::cube inputAsCube(input.memptr(), inRowSize, inColSize, depth);
+  arma::cube outputAsCube(output.memptr(), outRowSize, outColSize, depth, false, true);
+
+  double scaleRow = (double) inRowSize / (double) outRowSize;
+  double scaleCol = (double) inColSize / (double) outColSize;
+
+  arma::mat22 coeffs;
+  for (size_t i = 0; i < outRowSize; i++)
+  {
+    for (size_t j = 0; j < outColSize; j++)
+    {
+      size_t rOrigin = (size_t) std::floor(i * scaleRow);
+      size_t cOrigin = (size_t) std::floor(j * scaleCol);
+
+      if (rOrigin > inRowSize - 2)
+        rOrigin = inRowSize - 2;
+      if (cOrigin > inColSize - 2)
+        cOrigin = inColSize - 2;
+
+      double deltaR = i * scaleRow - rOrigin;
+      double deltaC = j * scaleCol - cOrigin;
+      coeffs[0] = (1 - deltaR) * (1 - deltaC);
+      coeffs[1] = deltaR * (1 - deltaC);
+      coeffs[2] = (1 - deltaR) * deltaC;
+      coeffs[3] = deltaR * deltaC;
+
+      for (size_t k = 0; k < depth; k++)
+      {
+        outputAsCube(i, j, k) = arma::accu(inputAsCube.slice(k).submat(
+                rOrigin, cOrigin, rOrigin + 1, cOrigin + 1) % coeffs);
+      }
+    }
+  }
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void BilinearInterpolation<InputDataType, OutputDataType>::Backward(
+    const arma::Mat<eT>&& /*input*/,
+    arma::Mat<eT>&& gradient,
+    arma::Mat<eT>&& output)
+{
+  if (output.is_empty())
+    output.set_size(inRowSize * inColSize * depth, 1);
+
+  assert(output.n_rows == inRowSize * inColSize * depth);
+  assert(output.n_cols == 1);
+  assert(outRowSize >= 2);
+  assert(outColSize >= 2);
+
+  arma::cube gradientAsCube(gradient.memptr(), outRowSize, outColSize, depth);
+  arma::cube outputAsCube(output.memptr(), inRowSize, inColSize, depth, false, true);
+
+  if (gradient.n_elem == output.n_elem)
+  {
+    outputAsCube = gradientAsCube;
+  }
+  else
+  {
+    double scaleRow = (double)(outRowSize) / inRowSize;
+    double scaleCol = (double)(outColSize) / inColSize;
+
+    arma::mat22 coeffs;
+    for (size_t i = 0; i < inRowSize; i++)
+    {
+      for (size_t j = 0; j < inColSize; j++)
+      {
+        size_t rOrigin = (size_t) std::floor(i * scaleRow);
+        size_t cOrigin = (size_t) std::floor(j * scaleCol);
+
+        if (rOrigin > outRowSize - 2)
+          rOrigin = outRowSize - 2;
+        if (cOrigin > outColSize - 2)
+          cOrigin = outColSize - 2;
+
+        double deltaR = i * scaleRow - rOrigin;
```

Comment:
Looks like we could apply the same optimizations as done for the forward pass here.